design/disaster-recovery-snapshots.md

```diff
@@ -0,0 +1,128 @@
+
+1. Improving application deployment process to avoid the `field is immutable` errors.
+1. Performing incremental backups of docker images.
+
+## Background
+
+Velero allows a number of ways to include and exclude resources in backups.
+In current implememntation, backup is created with the `includedNamespaces` key set to include all namespaces where application is deployed.
+All Admin Console resources have label `velero.io/exclude-from-backup: "true"`, which is recognized by Velero to exclude these resources.
+
+
+Disaster recovery snapshots must include Admin Console data, and potentially all resources.
+Due to this requirement, `velero.io/exclude-from-backup` label can no longer be used.
+The `includedNamespaces` filter also cannot be used because Admin Console and application can be deployed into the same namespace, which will cause problems during the restore process.
+
+## High-Level Design
+
+Velero backup spec supports the `labelSelector` field.
+All resources, application and Admin Console, can be labeled for inclusion in snapshots.
+Similarly, restore spec supports the same `labelSelector` field to select specific resources to restore.
+Using different `labelSelector` values allows to restore application and Admin Console separately.
+
+## Detailed Design
+
+### Backup
+
+- All resources (Admin Console and application) will have an additional label:
+    - `kots.io/backup: velero`
+- All application resources will have two additional labels:
+    - `kots.io/backup: velero`
+    - `kots.io/app-slug: <slug>`
+- Backup will be created with `labelSelector` set as in the example below:
+
+        apiVersion: velero.io/v1
+        kind: Backup
+        spec:
+          labelSelector:
+            matchLabels:
+              kots.io/backup: velero
+
+### Restore
+
+- Admin Console restore will be created with `labelSelector` set as in the example below:
+
+        apiVersion: velero.io/v1
+        kind: Restore
+        spec:
+          labelSelector:
+            matchLabels:
+              kots.io/kotsadm: "true"
+
+- Application restore will be created with `labelSelector` set as in the example below:
+
+        apiVersion: velero.io/v1
+        kind: Restore
+        spec:
+          labelSelector:
+            matchLabels:
+              kots.io/app-slug: <slug>
+
+- Restore can be performed using Velero CLI:
+
+        velero restore create --from-backup <backup name> -l kots.io/backup=velero
+
+        velero restore create --from-backup <backup name> -l kots.io/app-slug=<slug>
+
+- `kots install --from-backup` command will be implemented as a wrapper for the above commands
+    - `kots restore` is a potential alternative, but the UI-based approached discussed next will also need a `kots install` command.
+
+
+### Potential blockers
+
+When adding new labels, the next deployment can potentially fail with the `field is immutable` errors.
+A workaround is to delete the asset that causes the error prior to deploying the new version.
```

Comment:
Do you mean that when we add the `kots.io/backup` labels, we'll have to delete objects and recreate them? Is that risky? It must cause downtime.